File: colocateexceptions/osgi/src/impl/com/extensiblejava/calculator/MinimumPaymentScheduleCalculator.java
package com.extensiblejava.calculator;

import java.math.*;
import com.extensiblejava.loan.*;
//import com.extensiblejava.loan.impl.*;

public class MinimumPaymentScheduleCalculator implements LoanCalculator {
	/* private BigDecimal presentValue;
	private BigDecimal rate;
	private int term; */
	private PaymentFactory paymentFactory;
	//private BigDecimal cumulativePrincipal = new BigDecimal("0");
	//private BigDecimal cumulativeInterest = new BigDecimal("0");

	public MinimumPaymentScheduleCalculator(PaymentFactory paymentFactory) {
		this.paymentFactory = paymentFactory;
	}

	public Loan calculateLoan(BigDecimal presentValue, BigDecimal rate, int term) throws CalculationException {
		BigDecimal cumulativePrincipal = new BigDecimal("0");
		BigDecimal cumulativeInterest = new BigDecimal("0");
		try {
			PaymentSchedule paymentSchedule = this.paymentFactory.createPaymentSchedule();
			BigDecimal adjustedRate = rate.divide(new BigDecimal("1200"), 2, BigDecimal.ROUND_HALF_UP);
			MonthlyPaymentCalculator paymentCalculator = new MonthlyPaymentCalculator();
			BigDecimal monthlyPayment = paymentCalculator.calculatePayment(presentValue, rate, term);
			BigDecimal loanBalance = new BigDecimal(presentValue.toString());
			while (loanBalance.doubleValue() > monthlyPayment.doubleValue()) {
				BigDecimal interest = loanBalance.multiply(adjustedRate);
				interest = interest.setScale(2, BigDecimal.ROUND_HALF_UP);
				BigDecimal principal = monthlyPayment.subtract(interest);
				principal = principal.setScale(2, BigDecimal.ROUND_HALF_UP);
				//Payment payment = new PaymentImpl(principal, interest);
				Payment payment = this.paymentFactory.createPayment(principal, interest); 
				paymentSchedule.addPayment(payment);

				cumulativeInterest = cumulativeInterest.add(interest).setScale(2, BigDecimal.ROUND_HALF_UP);
				cumulativePrincipal = cumulativePrincipal.add(principal).setScale(2, BigDecimal.ROUND_HALF_UP);
				loanBalance = loanBalance.subtract(principal);
			}

			BigDecimal interest = loanBalance.multiply(adjustedRate).setScale(2, BigDecimal.ROUND_HALF_UP);
			BigDecimal principal = loanBalance.setScale(2, BigDecimal.ROUND_HALF_UP);
			cumulativeInterest = cumulativeInterest.add(interest).setScale(2, BigDecimal.ROUND_HALF_UP);
			cumulativePrincipal = cumulativePrincipal.add(principal).setScale(2, BigDecimal.ROUND_HALF_UP);
			//Payment payment = new PaymentImpl(principal, interest);
			Payment payment = this.paymentFactory.createPayment(principal, interest);
			paymentSchedule.addPayment(payment);
			//return paymentSchedule;
			return this.paymentFactory.createLoan(paymentSchedule, cumulativeInterest, cumulativePrincipal);
		} catch (Exception e) {
			throw new CalculationException(e);
		}
		//PaymentSchedule paymentSchedule = new PaymentScheduleImpl();
		
	}

	//public BigDecimal getCumulativeInterest() { return this.cumulativeInterest; }
	//public BigDecimal getCumulativePrincipal() { return this.cumulativePrincipal; }
}

File: colocateexceptions/osgi/scala/src/MinimumPaymentScheduleCalculator.scala
package com.extensiblejava.calculator.scala

import com.extensiblejava.loan._
//import MonthlyPaymentCalculator._

class MinimumPaymentScheduleCalculator(paymentFactory:PaymentFactory) extends LoanCalculator {
	def calculateLoan(value:java.math.BigDecimal, rate:java.math.BigDecimal, term:Int): Loan = {
	
		Console.println("---** IN SCALA CALCULATOR **---")
		val mc = new java.math.MathContext(2, java.math.RoundingMode.HALF_UP)
		val presentValue = BigDecimal(value)
		val presentRate = BigDecimal(rate).apply(mc)
		
		var cumulativePrincipal = BigDecimal("0.00")
		var cumulativeInterest = BigDecimal("0.00")
		
		val paymentSchedule = paymentFactory.createPaymentSchedule()
		val adjustedRate = (presentRate / (BigDecimal("1200")).setScale(2, BigDecimal.RoundingMode.HALF_UP))
		val calculator = new MonthlyPaymentCalculator()
		val monthlyPayment = calculator.calculatePayment(presentValue, presentRate, term)
		
		var loanBalance = BigDecimal(presentValue.bigDecimal)
		
		while (loanBalance.toDouble > monthlyPayment.toDouble) {
			val interest = (loanBalance.*(adjustedRate)).setScale(2, BigDecimal.RoundingMode.HALF_UP)
			val principal = (monthlyPayment.-(interest)).setScale(2, BigDecimal.RoundingMode.HALF_UP);
			val payment = paymentFactory.createPayment(principal.bigDecimal, interest.bigDecimal); 
			paymentSchedule.addPayment(payment);

			cumulativeInterest = cumulativeInterest.+(interest).setScale(2, BigDecimal.RoundingMode.HALF_UP);
			cumulativePrincipal = cumulativePrincipal.+(principal).setScale(2, BigDecimal.RoundingMode.HALF_UP);
			loanBalance = loanBalance.-(principal);
		}
		val interest = (loanBalance.*(adjustedRate)).setScale(2, BigDecimal.RoundingMode.HALF_UP)
		val principal = loanBalance.setScale(2, BigDecimal.RoundingMode.HALF_UP)
		cumulativeInterest = cumulativeInterest.+(interest).setScale(2, BigDecimal.RoundingMode.HALF_UP)
		cumulativePrincipal = cumulativePrincipal.+(principal).setScale(2, BigDecimal.RoundingMode.HALF_UP)
		val payment = paymentFactory.createPayment(principal.bigDecimal, interest.bigDecimal)
		paymentSchedule.addPayment(payment)
		
		val loan = paymentFactory.createLoan(paymentSchedule, cumulativeInterest.bigDecimal, cumulativePrincipal.bigDecimal)
		loan
		
	}
}